charms/interface_kubernetes_cni.py

```diff
@@ -0,0 +1,74 @@
+import hashlib
+import os
+
+
+class KubernetesCniProvides:
+    """Implements the Provides side of the kubernetes-cni interface."""
+
+    def __init__(self, charm, endpoint, default_cni):
+        self.charm = charm
+        self.endpoint = endpoint
+        self.default_cni = default_cni
+
+    @property
+    def cidr(self):
```

Comment:
Maybe this should provide an `ipaddress.IPv4Network` or `ipaddress.IPv6Network`
```suggestion
    def cidr(self) -> Optional[Union[IPv4Network, IPv6Network]]
```

Reply:
this is a bit of an API change for downstream users not expecting a `IPv4Network|IPv6Network` object

Reply:
CIDR can be something like `"192.168.0.0/16,fd00:c00b:1::/112"`, so I guess it would have to be `List[Union[IPv4Network, IPv6Network]]`

Normally we just pass this string straight into kube-apiserver, kube-controller-manager, and kube-proxy as-is. I'm open to parsing it here if you think it'd be cleaner that way.